docs/tech_specs/screens_by_alert/tech_spec.md

```diff
@@ -80,6 +64,37 @@ System.convert_time_unit(time2 - time1, :native, :second)
 
 </details>
 
+### Timers and TTLs
+
+In addition to the actual structure of the data being stored in the cache, we also need to carefully choose various timings and cache TTLs to properly keep that data up to date.
+
+TTLs in our case are also somewhat tricky, because some of them will be carried out internally by memcached, and others will need to be carried out by us in Elixir code.
+
+**Timers**
+| timer | interval (sec) | Why? |
+| - | - | - |
+| "self-refresh" job-runner GenServer | 70 | Job should give the slowest screen clients (e-ink, 30 sec refresh rate) time to retry after 1 missing request. |
+
+**TTLs**
+| value | Handled by | TTL (sec) | Why? |
+| - | - | - | - |
+| `screens_by_alert.*` cache items | memcached | 75 seconds | A key should expire shortly after a self-refresh runs without updating that key. |
+| screen IDs listed under a `screens_by_alert.*` key | application logic | 75 seconds | A screen ID should expire from the list shortly after a self-refresh runs without reaffirming the ID's place in the list. |
```

Comment:
> A screen ID should expire from the list shortly after a self-refresh runs without reaffirming the ID's place in the list.

Is this mechanism described in this doc? (Does it need to be?)

To help wrap my head around this, here's my scenario:  let's say an alert is bumped off a screen by a higher priority alert. So instead of having alerts A and B on Screen 1, it now has alerts A and C. The screen declares to the cache it has different alerts than it did before. Alert A still has Screen 1 in its list, Alert C now has Screen 1 in its list, and B still has it in its list but with an old timestamp.

in clunky pseudo:
%{
  "screens_by_alert.A” => [{Screen1, current time}, {ScreenY, current time}]
  "screens_by_alert.B” => [{Screen1, time from 30s ago}, {ScreenX, current time}]
  "screens_by_alert.C” => [{Screen1, current time}]
  “screens_last_updated.1” => current time
}

Kinda like that. Each alert is still getting reported by at least 1 screen, so we never hit the TTL for "screens_by_alert.*"

QUESTION:  What mechanism removes Screen1 from AlertB’s screen list?

Reply:
Every time the list under an alert ID is updated, we also _remove_ any list elements whose timestamps have expired. In this instance, ScreenX is still showing alertB, so when its client next fetches data, the resulting cache update will put a new `{ScreenX, now}` tuple in the list under `screens_by_alert.B`, _and_ remove both the previous ScreenX entry and the Screen1 entry, since they're now expired.

In fact, we could also do this cleanup when Screenplay asks for cached data. Any time we `get` an alert ID cache key, it's an opportunity to do a subsequent `set` that clears expired screen IDs from the list.

**In the case of clients that refresh more frequently, like pre-fare's 20s rate, the list may occasionally contain duplicate screen IDs with non-expired timestamps. That's ok—it's simple to remove duplicates from the list in code.

- - -

Other ways the data reaches "eventual accuracy":

If the only screen(s) still showing an alert have broken clients for some reason, then the self-refresh job will cover them in keeping the list up to date.

If an alert is no longer being shown by any screens at all, but still exists in the API data, then the `screens_by_alert.<id>` cache key will eventually expire due to lack of updates.

If an alert is no longer being shown by any screens at all, and has been cleared from the API data, then Screenplay will know about that earlier since it also checks the API regularly. The `screens_by_alert.<id>` cache key will eventually expire due to lack of updates.